### PR TITLE
[SPARK-41483][CORE] Last metrics system report should have a timeout, avoid to lead shutdown hook timeout

### PR DESCRIPTION
**Pull request by packyan**
### What changes were proposed in this pull request?

In the stop method of SparkContext, a timeout limit is added to metricsSystem#report. When it is not completed for more than 1s, continue with the subsequent steps.

### Why are the changes needed?

When the user does not explicitly call the stop method of SparkContext, the stop will be executed by ShutdownHook, and the Hadoop shutdown hookmanager used by Spark needs to complete the hook entry within 10s, so when the metricsSystem takes a long time to complete the report (for example, the sink occurs time out), other hooks may be blocked and cannot be executed successfully.

For example, the unregister operation in ApplicationMaster may fail to execute because the metrics report costs too much time, so the Spark job has been successfully completed, but Yarn still considers it a failed job.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

test manually, due to the difficulty of constructing the timeout execution of metrics#report. Unit test want help!

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
@packyan mind creating a JIRA and linking it to PR title? See also https://spark.apache.org/contributing.html

**Comment by packyan** [comment]
> @packyan mind creating a JIRA and linking it to PR title? See also https://spark.apache.org/contributing.html

Sorry, I will do it later.

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 1]
```diff
@@ -2136,7 +2138,10 @@ class SparkContext(config: SparkConf) extends Logging {
     }
     if (env != null) {
       Utils.tryLogNonFatalError {
-        env.metricsSystem.report()
+        val metricsReportFuture = Future {
+          env.metricsSystem.report()
+        } (SparkContext.executionContext)
+        ThreadUtils.awaitResult(metricsReportFuture, Duration.create("1s"))
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/SparkContext.scala:2144]: 1s seems to be a bit aggressive. How long would it take normally?

Comment (packyan) [reply core/src/main/scala/org/apache/spark/SparkContext.scala:2144]: I think 1 to 2 seconds is reasonable.  The last metrics report is not necessary, we must reserve enough time for the application master hook to unregister and clean up temporary files.

What do you think?

**Comment by packyan** [comment]
Any one can give me some suggestions?

**Pull request closed without merging**